homeassistant/components/octoprint/config_flow.py

```diff
@@ -45,43 +45,77 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
     def __init__(self):
         """Handle a config flow for OctoPrint."""
         self.discovery_schema = None
-        self.user_input = {}
+        self._user_input = {}
 
     async def async_step_user(self, user_input=None):
         """Handle the initial step."""
+        # When coming back from the progress steps, the user_input is stored in the
+        # instance variable instead of being passed in
+        if not user_input and len(self._user_input) > 0:
+            user_input = self._user_input
+
         if user_input is None:
             data = self.discovery_schema or _schema_with_defaults()
             return self.async_show_form(step_id="user", data_schema=data)
 
-        self.user_input = user_input
         if CONF_API_KEY in user_input:
-            return await self.async_step_finish(user_input)
+            try:
+                return await self._finish_config(user_input)
+            except data_entry_flow.AbortFlow as err:
+                raise err from None
+            except CannotConnect:
+                return self.async_show_form(
```

Comment:
Just store the error dict in a variable and show the form after all the except blocks if there are errors.